src/system/ExtensionsModule/Api/VariableApi.php

```diff
@@ -0,0 +1,255 @@
+     *
+     * @return mixed - extension variable value
+     */
+    public function get($extensionName, $variableName, $default = false)
+    {
+        if (empty($extensionName) || !is_string($extensionName) || empty($variableName) || !is_string($variableName)) {
+            throw new \InvalidArgumentException();
+        }
+        if (!$this->initFlag) {
+            $this->initExtensionVars();
+        }
+
+        if (isset($this->variables[$extensionName]) && array_key_exists($variableName, $this->variables[$extensionName])) {
+            return $this->variables[$extensionName][$variableName];
+        }
+
+        return $default;
+    }
+
+    /**
+     * Get all the variables for an extension.
+     *
+     * @param $extensionName
+     * @return array
+     */
+    public function getAll($extensionName)
+    {
+        if (empty($extensionName) || !is_string($extensionName)) {
+            throw new \InvalidArgumentException();
+        }
+        if (!$this->initFlag) {
+            $this->initExtensionVars();
+        }
+
+        return isset($this->variables[$extensionName]) ? $this->variables[$extensionName] : [];
+    }
+
+    /**
+     * Set an extension variable.
+     *
+     * @param string $extensionName The name of the extension.
+     * @param string $variableName The name of the variable.
+     * @param string $value The value of the variable.
+     *
+     * @return boolean True if successful, false otherwise.
+     */
+    public function set($extensionName, $variableName, $value = '')
+    {
+        if (empty($extensionName) || !is_string($extensionName) || empty($variableName) || !is_string($variableName)) {
+            throw new \InvalidArgumentException();
+        }
+
+        $entities = $this->repository->findBy(array('modname' => $extensionName, 'name' => $variableName));
+        if (count($entities) > 1 || count($entities) == 0) {
+            foreach ($entities as $entity) {
+                // possible duplicates exist. remove all (refs #2385)
+                $this->repository->remove($entity);
+            }
+            $entity = new ExtensionVarEntity();
+            $entity->setModname($extensionName);
+            $entity->setName($variableName);
+        } else {
+            $entity = $entities[0];
+        }
+        $entity->setValue($value);
+        $this->repository->persistAndFlush($entity);
+        $this->variables[$extensionName][$variableName] = $value;
+
+        return true;
+    }
+
+    /**
+     * The setAll method sets multiple extension variables.
+     *
+     * @param string $extensionName The name of the extension.
+     * @param array $vars An associative array of varnames/varvalues.
+     *
+     * @return boolean True if successful, false otherwise.
+     */
+    public function setAll($extensionName, array $vars)
+    {
+        $ok = true;
+        foreach ($vars as $var => $value) {
+            $ok = $ok && $this->set($extensionName, $var, $value);
```

Comment:
This does change the logic: Your approach would also delete all other variables of the extension. The current approach does not touch variables not present in the $vars array.